lib/hanami/early_hints.rb

```diff
@@ -0,0 +1,129 @@
+# frozen_string_literal: true
+
+module Hanami
+  # HTTP/2 Early Hints Rack middleware
+  #
+  # It sends extra responses **before** the main reponse is sent.
+  # These extra responses are HTTP/2 Early Hints (103).
+  # They specify the web assets (javascripts, stylesheets, etc..) to be "pushed",
+  # so modern browsers pre-fetch them in parallel with the main HTTP response.
+  #
+  # @see https://tools.ietf.org/html/draft-ietf-httpbis-early-hints-05
+  #
+  # @since x.x.x
+  # @api private
+  class EarlyHints
+    # @since x.x.x
+    # @api private
+    class NotSupportedByServerError < ::StandardError
+      # @since x.x.x
+      # @api private
+      def initialize
+        super("Current Ruby server doesn't support Early Hints.\nPlease make sure to use Puma with Early Hints enabled.")
```

Comment:
Why not better say
>Please make sure to use a web server with Early Hints enabled
or
>Please make sure to use a web server with support for Early Hints.

Reply:
Yes, just Puma for now. That's why my suggestion is really specific. As long other vendors will add support, we can change this message.

Reply:
@AlfonsoUceda I rephrased this: https://github.com/hanami/hanami/pull/905/commits/44d25a35896b43c39326f87dbf979e66c7bdf61f